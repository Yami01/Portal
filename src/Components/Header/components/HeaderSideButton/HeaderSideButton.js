/**
 *
 * HeaderSideButton
 * @format
 * @flow
 *
 */

import React, { memo } from 'react'
import type { Node, AbstractComponent } from 'react'
import { View } from 'react-native'
import { TouchableRipple } from 'react-native-paper'

import {
  BADGE_SIZE,
  BADGE_TEST_ID,
  DOT_TEST_ID,
  LINE_HEIGHT,
  TOUCHABLE_WRAPPER_TEST_ID,
} from './config'

import HeaderSideButtonStyles from './styles'
import type { PropsType } from './types'
import { Colors } from '@/Theme/Variables'
import Badge from '@/Components/Badge'
import Icon from '@/Components/Icon'
import Text from '@/Components/Text'

const HeaderSideButton: AbstractComponent<PropsType> = memo(
  (props: PropsType): Node => {
    const {
      rounded,
      iconGroup,
      iconName,
      iconFillColor,
      shouldShowLabel,
      label,
      badgeNumber,
      isShowDot,
      dotBorderColor,
      roundedButtonBackgroundColor = Colors.white,
      testID,
      onPress,
    } = props

    const containerStyle = [
      HeaderSideButtonStyles.container,
      rounded && [
        HeaderSideButtonStyles.containerRounded,
        {
          backgroundColor: roundedButtonBackgroundColor,
        },
      ],
    ]

    const badgeStyle = [
      HeaderSideButtonStyles.absoluteDecoration,
      HeaderSideButtonStyles.badge,
    ]

    const dotStyle = [
      HeaderSideButtonStyles.absoluteDecoration,
      HeaderSideButtonStyles.dot,
      {
        borderColor: dotBorderColor,
        backgroundColor: iconFillColor,
      },
    ]

    return (
      <View>
        <TouchableRipple delayPressIn={0} onPress={onPress} {...testID}>
          <View style={containerStyle} testID={TOUCHABLE_WRAPPER_TEST_ID}>
            {shouldShowLabel && !!label && (
              <Text style={HeaderSideButtonStyles.label}>{label}</Text>
            )}
            <Icon
              fillColor={iconFillColor}
              group={iconGroup}
              name={iconName}
              width={LINE_HEIGHT}
              height={LINE_HEIGHT}
            />
          </View>
        </TouchableRipple>
        {!isShowDot && !!badgeNumber && (
          <Badge
            extraStyle={badgeStyle}
            size={BADGE_SIZE}
            testID={BADGE_TEST_ID}
          >
            {badgeNumber}
          </Badge>
        )}
        {isShowDot && <View style={dotStyle} testID={DOT_TEST_ID} />}
      </View>
    )
  },
)

HeaderSideButton.displayName = 'HeaderSideButton'

export default HeaderSideButton
