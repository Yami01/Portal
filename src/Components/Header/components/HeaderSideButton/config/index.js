/**
 *
 * @format
 * @flow
 *
 */

export const ROUNDED_BUTTON_PADDING = 8
export const LINE_HEIGHT = 24
export const BADGE_SIZE = 15
export const TOUCHABLE_WRAPPER_TEST_ID = 'touchableWrapperTestID'
export const BADGE_TEST_ID = 'badgeTestID'
export const DOT_TEST_ID = 'dotTestID'
