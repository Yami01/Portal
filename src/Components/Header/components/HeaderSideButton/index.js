/**
 *
 * @flow
 *
 */

export { default } from './HeaderSideButton'
