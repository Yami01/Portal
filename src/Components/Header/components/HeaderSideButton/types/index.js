/**
 *
 * @format
 * @flow
 *
 */
import type {
  ViewStyleProp,
  TextStyleProp,
} from 'react-native/Libraries/StyleSheet/StyleSheet'
import type { PressEvent } from 'react-native/Libraries/Types/CoreEventTypes'

export type PropsType = {
  rounded?: boolean,
  iconGroup: string,
  iconName: string,
  iconFillColor: string,
  label?: string,
  badgeNumber?: number,
  isShowDot?: boolean,
  dotBorderColor?: string,
  shouldShowLabel?: boolean,
  // Only available when rounded true. Default to WHITE.
  roundedButtonBackgroundColor?: string,
  testID?: string,
  onPress?: (event: PressEvent) => any,
}

export type HeaderSideButtonStyleType = {
  absoluteDecoration: ViewStyleProp,
  badge: ViewStyleProp,
  container: ViewStyleProp,
  containerRounded: ViewStyleProp,
  dot: ViewStyleProp,
  label: TextStyleProp,
}
