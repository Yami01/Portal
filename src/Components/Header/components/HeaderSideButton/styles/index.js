/**
 *
 * @format
 * @flow
 *
 */

import { StyleSheet } from 'react-native'
import type { HeaderSideButtonStyleType } from '../types'
import { LINE_HEIGHT, ROUNDED_BUTTON_PADDING } from '../config'
import { Colors } from '@/Theme/Variables'

const HeaderSideButtonStyles: HeaderSideButtonStyleType = StyleSheet.create({
  absoluteDecoration: {
    position: 'absolute',
    top: -3,
    right: -3,
    zIndex: 10,
  },
  badge: {
    borderWidth: 1,
    borderColor: Colors.white,
    backgroundColor: Colors.red_950,
  },
  container: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  containerRounded: {
    padding: ROUNDED_BUTTON_PADDING,
    margin: -ROUNDED_BUTTON_PADDING,
    borderRadius: (LINE_HEIGHT + 2 * ROUNDED_BUTTON_PADDING) / 2,
    elevation: 5,
    shadowColor: Colors.gray,
    shadowOffset: { width: 2, height: 3 },
    shadowOpacity: 0.15,
    shadowRadius: 2,
  },
  dot: {
    borderWidth: 2,
    backgroundColor: Colors.white,
    width: 12,
    height: 12,
    borderRadius: 6,
  },
  label: {
    color: Colors.white,
    fontSize: 12,
    marginRight: 5,
  },
})

export default HeaderSideButtonStyles
