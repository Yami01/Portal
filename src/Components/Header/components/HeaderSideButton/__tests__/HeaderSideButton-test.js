/**
 *
 * HeaderSideButton Test
 * @format
 * @flow
 *
 */

import React from 'react'
import renderer from 'react-test-renderer'
import { render, fireEvent } from '@testing-library/react-native'

import HeaderSideButton from '../HeaderSideButton'
import {
  BADGE_TEST_ID,
  DOT_TEST_ID,
  TOUCHABLE_WRAPPER_TEST_ID,
} from '../config'
import HeaderSideButtonStyles from '../styles'

describe('HeaderSideButton', () => {
  const label = 'Updated today'
  const badgeNumber = 1
  const dotBorderColor = '#098765'
  const testID = 'headerButtonTestID'
  const buttonBackgroundColor = '#135790'
  const props = {
    iconFillColor: '#fff',
    iconGroup: 'general',
    iconName: 'menu',
    testID,
  }

  test('render with icon only', () => {
    const tree = renderer.create(<HeaderSideButton {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('render with dot', () => {
    const tree = renderer
      .create(
        <HeaderSideButton
          {...props}
          isShowDot
          dotBorderColor={dotBorderColor}
        />,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('render with label and badge', () => {
    const tree = renderer
      .create(
        <HeaderSideButton
          {...props}
          shouldShowLabel
          label={label}
          badgeNumber={badgeNumber}
        />,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('is accessibility works', () => {
    const { getByTestId } = render(<HeaderSideButton {...props} />)

    expect(getByTestId(testID)).not.toBeNull()
  })

  test('is label not show if shouldShowLabel not true', () => {
    const { queryByText } = render(
      <HeaderSideButton {...props} label={label} />,
    )

    expect(queryByText(label)).toBeNull()
  })

  test('is label show if shouldShowLabel true', () => {
    const { getByText } = render(
      <HeaderSideButton {...props} shouldShowLabel label={label} />,
    )

    expect(getByText(label)).not.toBeNull()
  })

  test('can render if badgeNumber is 0', () => {
    const { queryByTestId } = render(
      <HeaderSideButton {...props} badgeNumber={0} />,
    )

    expect(queryByTestId(testID)).not.toBeNull()
    expect(queryByTestId(BADGE_TEST_ID)).toBeNull()
  })

  test('can render with badgeNumber', () => {
    const { queryByTestId, queryByText } = render(
      <HeaderSideButton {...props} badgeNumber={badgeNumber} />,
    )

    expect(queryByTestId(testID)).not.toBeNull()
    expect(queryByTestId(BADGE_TEST_ID)).not.toBeNull()
    expect(queryByText(badgeNumber.toString())).not.toBeNull()
  })

  test('can render with dot', () => {
    const { queryByTestId } = render(
      <HeaderSideButton {...props} isShowDot dotBorderColor={dotBorderColor} />,
    )

    const dot = queryByTestId(DOT_TEST_ID)

    expect(queryByTestId(testID)).not.toBeNull()
    expect(queryByTestId(BADGE_TEST_ID)).toBeNull()
    expect(dot).not.toBeNull()
    expect(dot).toHaveStyle({ borderColor: dotBorderColor })
  })

  test('can render dot only even badgeNumber provider', () => {
    const { queryByTestId } = render(
      <HeaderSideButton
        {...props}
        isShowDot
        dotBorderColor={dotBorderColor}
        badgeNumber={badgeNumber}
      />,
    )

    const dot = queryByTestId(DOT_TEST_ID)

    expect(queryByTestId(testID)).not.toBeNull()
    expect(queryByTestId(BADGE_TEST_ID)).toBeNull()
    expect(dot).not.toBeNull()
  })

  test('can render as rounded button', () => {
    const { queryByTestId } = render(
      <HeaderSideButton
        {...props}
        rounded
        roundedButtonBackgroundColor={buttonBackgroundColor}
      />,
    )

    const touchableWrapper = queryByTestId(TOUCHABLE_WRAPPER_TEST_ID)

    expect(touchableWrapper).not.toBeNull()
    expect(touchableWrapper).toHaveStyle(
      HeaderSideButtonStyles.containerRounded,
    )
    expect(touchableWrapper).toHaveStyle({
      backgroundColor: buttonBackgroundColor,
    })
  })

  test('is roundedButtonBackgroundColor not affecting if rounded is false', () => {
    const { queryByTestId } = render(
      <HeaderSideButton
        {...props}
        roundedButtonBackgroundColor={buttonBackgroundColor}
      />,
    )

    const touchableWrapper = queryByTestId(TOUCHABLE_WRAPPER_TEST_ID)

    expect(touchableWrapper).not.toBeNull()
    expect(touchableWrapper).not.toHaveStyle(
      HeaderSideButtonStyles.containerRounded,
    )
    expect(touchableWrapper).not.toHaveStyle({
      backgroundColor: buttonBackgroundColor,
    })
  })

  test('is pressable', () => {
    const onPressMock = jest.fn()

    const { getByTestId } = render(
      <HeaderSideButton {...props} onPress={onPressMock} />,
    )

    const button = getByTestId(testID)
    fireEvent.press(button)
    expect(onPressMock).toHaveBeenCalled()
  })
})
