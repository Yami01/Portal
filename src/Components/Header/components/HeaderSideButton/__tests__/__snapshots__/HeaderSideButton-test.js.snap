// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderSideButton render with dot 1`] = `
<View>
  <View
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        false,
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          },
          undefined,
        ]
      }
      testID="touchableWrapperTestID"
    >
      <View
        style={
          Array [
            undefined,
            undefined,
          ]
        }
      />
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "position": "absolute",
          "right": -3,
          "top": -3,
          "zIndex": 10,
        },
        Object {
          "backgroundColor": "#ffffff",
          "borderRadius": 6,
          "borderWidth": 2,
          "height": 12,
          "width": 12,
        },
        Object {
          "backgroundColor": "#fff",
          "borderColor": "#098765",
        },
      ]
    }
    testID="dotTestID"
  />
</View>
`;

exports[`HeaderSideButton render with icon only 1`] = `
<View>
  <View
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        false,
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          },
          undefined,
        ]
      }
      testID="touchableWrapperTestID"
    >
      <View
        style={
          Array [
            undefined,
            undefined,
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`HeaderSideButton render with label and badge 1`] = `
<View>
  <View
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        false,
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          },
          undefined,
        ]
      }
      testID="touchableWrapperTestID"
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "System",
              "fontWeight": "400",
            },
            Object {
              "color": "#ffffff",
              "fontSize": 12,
              "marginRight": 5,
            },
          ]
        }
      >
        Updated today
      </Text>
      <View
        style={
          Array [
            undefined,
            undefined,
          ]
        }
      />
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "borderRadius": 7.5,
          "height": 15,
          "width": 15,
        },
        Object {
          "alignItems": "center",
          "backgroundColor": "#E14032",
          "justifyContent": "center",
        },
        Array [
          Object {
            "position": "absolute",
            "right": -3,
            "top": -3,
            "zIndex": 10,
          },
          Object {
            "backgroundColor": "#e10c0a",
            "borderColor": "#ffffff",
            "borderWidth": 1,
          },
        ],
      ]
    }
    testID="badgeTestID"
  >
    <Text
      style={
        Array [
          Object {
            "fontFamily": "System",
            "fontWeight": "400",
          },
          Array [
            Object {
              "color": "#ffffff",
              "fontSize": 10,
            },
            undefined,
          ],
        ]
      }
    >
      1
    </Text>
  </View>
</View>
`;
